graphprot/NeuralNet.py

```diff
@@ -53,61 +55,142 @@ def __init__(self, database, Net,
-                    epoch, loss, val_loss, t))
+                if acc is not None :
+                    print('Epoch [%04d] : train loss %e | train accuracy %1.4e | valid loss %e | val accuracy %1.4e | time %1.2e sec.' % (
+                        epoch, loss, acc, val_loss, val_acc, t))
+                else :
+                    print('Epoch [%04d] : train loss %e | valid loss %e | time %1.2e sec.' % (
+                        epoch, loss, val_loss, t))
+
             else:
                 t = time() - t0
-                print('Epoch [%04d] : train loss %e | time %1.2e sec.' % (
+                if acc is not None :
+                    print('Epoch [%04d] : train loss %e | accuracy %1.4e | time %1.2e sec. |' % (
+                    epoch, loss, acc, t))
+                else :
+                    print('Epoch [%04d] : train loss %e | time %1.2e sec. |' % (
                     epoch, loss, t))
 
+
+    def calcAccuracy(self, prediction, target, reduce=True):
+
+        overlap = (prediction.argmax(dim=1)==target).sum()
+        if reduce:
+            return overlap/float(target.size()[-1])
+
+        return overlap
+
+
     def eval(self, loader):
 
-        self.model.eval()
+        with self.model.eval():
 
-        loss_func, loss_val = self.loss, 0
-        out = []
-        for data in loader:
-            data = data.to(self.device)
-            pred = self.model(data).reshape(-1)
-            loss_val += loss_func(pred, data.y)
-            out += pred.reshape(-1).tolist()
-        return out, loss_val
+            loss_func, loss_val = self.loss, 0
+            out = []
+            acc = []
+            for data in loader:
+                data = data.to(self.device)
+                pred = self.model(data)
+
+                if self.task == 'class':
+                    pred = F.softmax(pred, dim=1)
+                    data.y = torch.tensor([self.classes_idx[int(x)] for x in data.y])
+                    acc.append(self.calcAccuracy(pred, data.y))
+
+                else :
+                    out = out.reshape(-1)
+                    acc = None
+
+                loss_val += loss_func(pred, data.y)
+                out += pred.reshape(-1).tolist()
+
+        if self.task == 'class':
+            return out, torch.mean(torch.stack(acc)), loss_val
 
+        else :
+            return out, acc, loss_val
+            
+            
     def _epoch(self, epoch):
 
         running_loss = 0
         for data in self.train_loader:
             data = data.to(self.device)
             self.optimizer.zero_grad()
-            out = self.model(data).reshape(-1)
+            out = self.model(data)
+
+            if self.task == 'class':
+                out = F.softmax(out, dim=1)
+                data.y = torch.tensor([self.classes_idx[int(x)] for x in data.y])
+                acc.append(self.calcAccuracy(out, data.y))
+
+            else :
+                out = out.reshape(-1)
+                acc = None
+
```

Comment:
you could encapsulate that in a function so that its's not duplicated in epoch and eval